clang/lib/Sema/SemaOpenACC.cpp

```diff
@@ -37,6 +37,32 @@ bool diagnoseConstructAppertainment(SemaOpenACC &S, OpenACCDirectiveKind K,
   }
   return false;
 }
+
+bool doesClauseApplyToDirective(OpenACCDirectiveKind DirectiveKind,
+                                OpenACCClauseKind ClauseKind) {
+  switch (ClauseKind) {
+    // FIXME: For each clause as we implement them, we can add the
+    // 'legalization' list here.
+  default:
+    // Do nothing so we can go to the 'unimplemented' diagnostic instead.
+    return true;
+  }
+  llvm_unreachable("Invalid clause kind");
+}
+
+/// Destruct and deallocate any clauses that aren't going to be used because
+/// they don't have a Construct to attach to.
+void DestroyUnusedClauses(ASTContext &Ctx,
```

Comment:
Why do you need to delete unused clauses? Maybe just keep them unused and just ignore?

Reply:
It just seemed like a bad idea to 'leak' clauses that aren't added to the AST.  Depending on how many 'bad' constructs there end up being, this could presumably be an unfortunate amount of memory.

Reply:
Not sure this is a real issue. I think it is enough just to report the error (or warning) and just ignore the clauses in the codegen.

Reply:
Well, there WOULD be no 'ignoring' in codegen, as these clauses won't be added to the AST at all.  They'd be created, then just 'forgotten' if we don't add them to a construct.

Reply:
I discussed this with Aaron who confirmed that because we're allocating in the ASTContext::Allocate pool that we're fine to skip destruction, so I'm just going to do that.